app/src/components/clients.js

```diff
@@ -0,0 +1,40 @@
+const config = require('config');
+
+const KeyCloakServiceClientManager = require('./keyCloakServiceClientMgr');
+const RealmAdminService = require('./realmAdminSvc');
+
+const clients = {
+
+  /**
+ * @function getClientsFromEnv
+ * Utility function to call the KC service to get service clients for a KC realm
+ * @param {string} kcEnv The KC env
+ * @returns {object[]} An array of service clients
+ */
+  getClientsFromEnv: async (kcEnv) => {
+
+    const realmKey = `serviceClient.keycloak.${kcEnv}`;
+    const {
+      endpoint: realmBaseUrl,
+      username: clientId,
+      password: clientSecret,
+      realm: realmId
+    } = config.get(realmKey);
+
+    const realmSvc = new RealmAdminService({ realmBaseUrl, clientId, clientSecret, realmId });
+    const kcScMgr = new KeyCloakServiceClientManager(realmSvc);
+
+    var clients = kcScMgr.fetchAllClients();
+    clients.then(result => {
+      // add a keycloak environment label to each service client object in the array
+      result.forEach(function (cl) {
+        cl.environment = kcEnv;
+      });
+    });
+
+    return clients;
```

Comment:
The way it is written right now, this function will return clients before the internal forEach is fully executed as clients itself will be an unresolved promise. Since this getClientsFromEnv function is already designated as async, you should instead do something like the following:

```
const clients = await kcScMgr.fetchAllClients();
clients.forEach(cl => cl.environment = kcEnv);

return clients;
```
This will make it so that all clients must have been resolved first before proceeding with the insertion of environment attributes synchronously and returning.

Reply:
thank you! i was trying to figure that one out. i see how that works now